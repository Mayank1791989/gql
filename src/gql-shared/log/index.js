/* @flow */
export { default } from './log';
export type { LogListener } from './log.js';
