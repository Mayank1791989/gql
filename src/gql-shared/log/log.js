/* @flow */
import Emitter, { type EmitterSubscription } from 'gql-shared/emitter';

const LOG_LEVELS = Object.freeze({
  debug: 'debug',
  info: 'info',
  error: 'error',
});
type LogLevel = $Values<typeof LOG_LEVELS>;

class Logger {
  _timeKeys = new Map();
  _emitter: any;
  _name: string;

  constructor(name, emitter) {
    this._emitter = emitter;
    this._name = name;
  }

  debug(...args: Array<mixed>) {
    this._emit(LOG_LEVELS.debug, ...args);
  }

  info(...args: Array<mixed>) {
    this._emit(LOG_LEVELS.info, ...args);
  }

  error(...args: Array<mixed>) {
    this._emit(LOG_LEVELS.info, ...args);
  }

  time(key: string) {
    if (this._timeKeys.has(key)) {
      console.warn(`Timer '${key}' already exists.`);
      return;
    }
    this._timeKeys.set(key, new Date());
  }

  timeEnd(key: string) {
    const startTime = this._timeKeys.get(key);
    // console.log(startTime);
    if (!startTime) {
      console.warn(`Timer '${key}' does not exist.`);
      return;
    }

    const timeTaken = new Date() - startTime;
    this._timeKeys.delete(key);
    this._emit(LOG_LEVELS.debug, `${key}: ${timeTaken}ms`);
  }

  _emit(level: LogLevel, ...args) {
    this._emitter.emit(level, {
      level,
      name: this._name,
      args,
    });
  }
}

export type LogListener = ({
  +level: LogLevel,
  +name: string,
  +args: Array<mixed>,
}) => void;

export default {
  _emitter: new Emitter(),

  onLog(listener: LogListener): EmitterSubscription {
    const subscriptions = Object.values(LOG_LEVELS).map(eventType => {
      return this._emitter.addListener(eventType, listener);
    });

    return {
      remove() {
        subscriptions.map(subscription => subscription.remove());
      },
    };
  },

  getLogger(name: string): Logger {
    return new Logger(name, this._emitter);
  },
};
