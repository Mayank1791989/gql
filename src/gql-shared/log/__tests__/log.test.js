/* @flow */
import log from '../log';

test('works', () => {
  const logger = log.getLogger('test');
  const mockOnLogHandler = jest.fn();
  log.onLog(mockOnLogHandler);

  logger.debug('test', 'mayank');
  logger.error('test', 'mayank');
  logger.info('info message');

  logger.time('test');
  logger.timeEnd('test');

  expect(mockOnLogHandler.mock.calls).toMatchSnapshot();
});
