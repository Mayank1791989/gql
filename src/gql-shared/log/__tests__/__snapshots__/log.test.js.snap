// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Array [
  Array [
    Object {
      "args": Array [
        "test",
        "mayank",
      ],
      "level": "debug",
      "name": "test",
    },
  ],
  Array [
    Object {
      "args": Array [
        "test",
        "mayank",
      ],
      "level": "info",
      "name": "test",
    },
  ],
  Array [
    Object {
      "args": Array [
        "info message",
      ],
      "level": "info",
      "name": "test",
    },
  ],
  Array [
    Object {
      "args": Array [
        "test: 0ms",
      ],
      "level": "debug",
      "name": "test",
    },
  ],
]
`;
